File: dotNET/PFProSDK/src/PayPal/Payments/DataObjects/ClientInfo.cs
#region "Imports"
using System;
using System.Collections;
using PayPal.Payments.Common.Utility;
#endregion

namespace PayPal.Payments.DataObjects
{
	/// <summary>
	/// This class is used to store the 
	/// Payflow Client related properties
	/// </summary>
	public sealed class ClientInfo : BaseRequestDataObject
	{
		private Hashtable mClientInfoHash = null;
		/// <summary>
		/// Constructor
		/// </summary>
		public ClientInfo()
		{
			mClientInfoHash = new Hashtable();
		}

		#region "Properties and getter, setter Methods"

		#region "Properties"
		/// <summary>
		/// Gets the client info hash table
		/// </summary>
		internal Hashtable ClientInfoHash
		{
			get { return mClientInfoHash;}
		}

		/// <summary>
		/// Gets client version
		/// </summary>
		/// <remarks>
		/// <para>Client version.</para>
		/// <para>Maps to HTTP Header:</para>
		/// <code>PAYFLOW-CLIENT-VERSION</code>
		/// </remarks>
		public String ClientVersion
		{
			get { return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_CLIENT_VERSION);}
		}
		/// <summary>
		/// Gets, sets OS architecture
		/// </summary>
		internal String OsArchitecture
		{
			get { return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_OS_ARCHITECTURE);}
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_OS_ARCHITECTURE, value); }
		}
		/// <summary>
		/// Gets, sets OS version
		/// </summary>
		internal String OsVersion
		{
			get{ return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_OS_VERSION);}
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_OS_VERSION, value); }
		}
		/// <summary>
		/// Gets, sets OS Name
		/// </summary>
		internal String OsName
		{
			get{ return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_OS_NAME);}
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_OS_NAME, value); }
		}
		/// <summary>
		/// Gets, sets Proxy
		/// </summary>
		internal String Proxy
		{
			get{ return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_PROXY);}
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_PROXY, value); }
		}
		/// <summary>
		/// Gets, sets runtime version
		/// </summary>
		internal String RunTimeVersion
		{
			get{ return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_RUNTIME_VERSION);}
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_RUNTIME_VERSION, value); }
		}
		/// <summary>
		/// Sets integration product
		/// </summary>
		/// <remarks>
		/// <para>Integration product.</para>
		/// <para>Maps to HTTP Header:</para>
		/// <code>PAYFLOW-INTEGRATION-PRODUCT</code>
		/// </remarks>
		public String IntegrationProduct
		{
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_PRODUCT, value); }
		}
		/// <summary>
		/// Sets integration version
		/// </summary>
		/// <remarks>
		/// <para>Integration product.</para>
		/// <para>Maps to HTTP Header:</para>
		/// <code>PAYFLOW-INTEGRATION-VERSION</code>
		/// </remarks>
		public String IntegrationVersion
		{
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_VERSION, value); }
		}
	
		/// <summary>
		/// Gets Client Type
		/// </summary>
		/// <remarks>
		/// <para>Client type.</para>
		/// <para>Maps to HTTP Header:</para>
		/// <code>PAYFLOW-CLIENT-TYPE</code>
		/// </remarks>
		public String ClientType
		{
			get { return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_CLIENT_TYPE);}
		}
		
		/// <summary>
		/// Sets Request Type
		/// </summary>
		/// <remarks>
		/// <para>Request type.</para>
		/// <para>Maps to HTTP Header:</para>
		/// <code>PAYFLOW-ASSEMBLY</code>
		/// </remarks>
		internal String RequestType
		{
			set { AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_ASSEMBLY, value);}
		}

		#endregion

		#region "Getters and setter"
		/// <summary>
		/// Sets client version
		/// </summary>
		/// <param name="Version">String value of client version</param>
		internal void SetClientVersion(String Version)
		{
			AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_CLIENT_VERSION, Version);
		}
		/// <summary>
		/// Gets integration product
		/// </summary>
		/// <returns>String value of integration product</returns>
		internal String GetIntegrationProduct()
		{
			return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_PRODUCT);
		}
		/// <summary>
		/// Gets integration version
		/// </summary>
		/// <returns>String value of integration version</returns>
		internal String GetIntegrationVersion()
		{
			return (String) GetHeaderFromHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_VERSION);
		}
		
		/// <summary>
		/// Sets client type
		/// </summary>
		/// <param name="ClientType">String value of Client Type</param>
		internal void SetClientType(String ClientType)
		{
				AddHeaderToHash(PayflowConstants.PAYFLOWHEADER_CLIENT_TYPE, ClientType);			
		}
		#endregion

		#endregion

		#region "Private and internal methods"
		/// <summary>
		/// Adds a header to the header hash table
		/// </summary>
		/// <param name="HeaderName">Header name</param>
		/// <param name="HeaderValue">Header value</param>
		internal void AddHeaderToHash(String HeaderName, Object HeaderValue)
		{
			// Null Header Names & Values are not allowed.
			// Empty Header Names are not allowed.
			if (HeaderName == null || HeaderName.Length == 0 ||HeaderValue == null)
			{
				return;
			}

			ClientInfoHeader CurrHeader
				= new ClientInfoHeader(HeaderName, HeaderValue);

			if (mClientInfoHash == null)
			{
				mClientInfoHash = new Hashtable();
			}

			if (mClientInfoHash.ContainsKey(HeaderName))
			{
				mClientInfoHash.Remove(HeaderName);
			}

			mClientInfoHash.Add(HeaderName, CurrHeader);
		}
		
		/// <summary>
		/// Gets a header value from hash
		/// </summary>
		/// <param name="HeaderName">Header name</param>
		/// <returns>Header value object</returns>
		internal Object GetHeaderFromHash(String HeaderName)
		{
			if (mClientInfoHash == null)
			{
				return null;
			}

			ClientInfoHeader CurrHeader = (ClientInfoHeader) mClientInfoHash[HeaderName];

			if (CurrHeader == null)
			{
				return null;
			}

			return CurrHeader.HeaderValue;


		}

		#endregion
	}
}

File: java/src/sdk/base/paypal/payflow/ClientInfo.java
package paypal.payflow;

import java.util.Hashtable;

/*
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * This class is used to store the Payflow Client related properties.
 */
public final class ClientInfo extends BaseRequestDataObject {
    private Hashtable clientInfoHash = null;

    /**
     * Constructor
     */
    public ClientInfo() {
        clientInfoHash = new Hashtable();
    }

    /**
     * Gets the client info hash table
     *
     * @return clientInfoHash HashTable
     */
    protected Hashtable getClientInfoHash() {
        return clientInfoHash;
    }

    /**
     * gets the clientVersion
     *
     * @return clientVersion
     */
    public String getClientVersion() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_CLIENT_VERSION);
    }

    /**
     * Gets the OS architecture
     *
     * @return osArchitecture String
     */
    protected String getOsArchitecture() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_OS_ARCHITECTURE);
    }

    /**
     * Sets the OS architecture
     *
     * @param value String
     */
    protected void setOsArchitecture(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_OS_ARCHITECTURE, value);
    }

    /**
     * gets the OS version
     *
     * @return osVersion String
     */
    protected String getOsVersion() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_OS_VERSION);
    }

    /**
     * sets the OsVersion
     *
     * @param value String
     */
    protected void setOsVersion(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_OS_VERSION, value);
    }

    /**
     * Gets the OS Name
     *
     * @return osName String
     */
    protected String getOsName() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_OS_NAME);
    }

    /**
     * sets the OsName
     *
     * @param value String
     */
    protected void setOsName(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_OS_NAME, value);
    }

    /**
     * Gets the Proxy header.
     *
     * @return proxy String
     */
    protected String getProxy() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_PROXY);
    }

    /**
     * sets the Proxy header
     *
     * @param value String
     */
    protected void setProxy(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_PROXY, value);
    }

    /**
     * Gets the Java runtime version
     *
     * @return runTimeVersion String
     */
    protected String getRunTimeVersion() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_RUNTIME_VERSION);
    }

    /**
     * Sets the Java runtime version
     *
     * @param value String
     */
    protected void setRunTimeVersion(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_RUNTIME_VERSION, value);
    }

    /**
     * Sets the integration product
     *
     * @param value String
     */
    public void setIntegrationProduct(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_PRODUCT, value);
    }

    /**
     * Sets the integration version
     *
     * @param value String
     */
    public void setIntegrationVersion(String value) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_VERSION, value);
    }

    /**
     * Gets Client Type
     *
     * @return ClientType String
     */
    public String getClientType() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_CLIENT_TYPE);
    }

    /**
     * Sets client version
     *
     * @param version String
     */
    protected void setClientVersion(String version) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_CLIENT_VERSION, version);
    }

    /**
     * Gets integration product
     *
     * @return IntegrationProduct String
     */
    protected String getIntegrationProduct() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_PRODUCT);
    }

    /**
     * Gets integration version
     *
     * @return IntegrationVersion String
     */
    protected String getIntegrationVersion() {
        return (String) getHeaderFromHash(PayflowConstants.PAYFLOWHEADER_INTEGRATION_VERSION);
    }

    /**
     * @param clientType String
     */
    protected void setClientType(String clientType) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_CLIENT_TYPE, clientType);
    }

    /**
     * payflowAssembly, the possible values are PayflowConstants.STRONG_ASSEMBLY or PayflowConstants.WEAK_ASSEMBLY
     *
     * @param payflowAssembly String
     */
    protected void setPayflowAssembly(String payflowAssembly) {
        addHeaderToHash(PayflowConstants.PAYFLOWHEADER_PAYFLOW_ASSEMBLY, payflowAssembly);
    }

    /**
     * Adds a header to the header hash table
     *
     * @param headerName  String
     * @param headerValue Object
     */
    protected void addHeaderToHash(String headerName, Object headerValue) {
        Logger.getInstance().log("paypal.payflow.ClientInfo.addHeaderToHash(String, Object): Entered", PayflowConstants.SEVERITY_DEBUG);
        // Null Header Names & Values are not allowed.
        // Empty Header Names are not allowed.
        if (!(headerName == null || headerName.length() == 0 || headerValue == null)) {

            ClientInfoHeader currHeader
                    = new ClientInfoHeader(headerName, headerValue);

            if (clientInfoHash == null) {
                clientInfoHash = new Hashtable();
            }

            if (clientInfoHash.containsKey(headerName)) {
                clientInfoHash.remove(headerName);
            }

            clientInfoHash.put(headerName, currHeader);
        }
        Logger.getInstance().log("paypal.payflow.ClientInfo.addHeaderToHash(String, Object): Exiting", PayflowConstants.SEVERITY_DEBUG);
    }

    /**
     * Gets a header value from hash
     *
     * @param headerName String
     * @return headerValue Object
     */
    protected Object getHeaderFromHash(String headerName) {
        Logger.getInstance().log("paypal.payflow.ClientInfo.getHeaderFromHash(String): Entered", PayflowConstants.SEVERITY_DEBUG);
        Object header = null;
        if (null != clientInfoHash) {
            ClientInfoHeader currHeader = (ClientInfoHeader) clientInfoHash.get(headerName);
            if (null != currHeader) {
                header = currHeader.getHeaderValue();
            }
        }
        Logger.getInstance().log("paypal.payflow.ClientInfo.getHeaderFromHash(String): Exiting", PayflowConstants.SEVERITY_DEBUG);
        return header;

    }

}